### PR TITLE
#838 Notes component

### LINKED ISSUE #838: JSON Forms widget -- multi-line text box
For entering "Notes" and other longer bits of information.

### PR DESCRIPTION
**Pull request by CarlosNZ**
Fix #838 

Okay, going to publish this now -- I think it still needs some work, but it'd be good to get some feedback, as I'm not sure what the best solution is to certain problems. Possibly we need to sit down for a short design meeting to agree on what the expected behaviour / look should be in various situations.

For now, I've taken the following approach:
- Created a new "Note" component which displays a single "Note", with "Text" and "Author" fields
- Used the existing "Array" component as a wrapper for multiple notes ("Array" is a fairly big component, seemed wasteful to duplicate a lot of the same behaviour in a new component)
- Added some additional "options" to "Array" component. They're specific to the "Notes", but I've tried to make them generic enough to be useful in other cases too (see code comment for specifics)
- Currently, only the most recent note is editable (and only if the author matches the current user, and it's less than a certain number of days old)

Things I'd like to discuss/get feedback on:
- What should the exact rules be for when a note is editable?
- The "header" for each note is currently just "Note #" (configured in `itemLabel` properety in uischema). Ideally the header would show the author/timestamp when it's collapsed, but this would require a bit of re-thinking to make it work nicely, so wondering what others think.

See inline comments for more info.

The "Notes" component is enabled in the "Patient" schema only, currently. Can add it to others as required.

### CODE REVIEW

File: client/packages/common/src/authentication/AuthContext.tsx
Comment (CarlosNZ) [inline client/packages/common/src/authentication/AuthContext.tsx:221]: An pre-existing oversight: the userId wasn't actually being returned in AuthContext.

File: client/packages/programs/src/JsonForms/components/Note.tsx
Comment (CarlosNZ) [inline client/packages/programs/src/JsonForms/components/Note.tsx:22]: Most options I've added are actually for the "Array" component -- these ones are basically the same as the "Text" input.

File: client/packages/programs/src/JsonForms/components/Note.tsx
Comment (CarlosNZ) [inline client/packages/programs/src/JsonForms/components/Note.tsx:30]: This component is mostly the same as the "Text" component, just with an extra line for "Author" field and a different display for when the note is not editable.

File: server/service/src/sync/program_schemas/patient_ui_schema.json
Comment (CarlosNZ) [inline server/service/src/sync/program_schemas/patient_ui_schema.json:457]: New "options" for "Array": `editRestrictions` and `reverse`

File: client/packages/programs/src/JsonForms/common/components/Array.tsx
Comment (clemens-msupply) [inline client/packages/programs/src/JsonForms/common/components/Array.tsx:435]: As above we should at least have a type definition for the options and cast the `any` type to this type definition, preferably with zod validation.

Maybe replace `data` and use Array.revert()?  I also thought about reverting the input `data` array before creating the array of Components but I am not sure if this would have some side effects...

File: client/packages/programs/src/JsonForms/components/Note.tsx
Comment (clemens-msupply) [inline client/packages/programs/src/JsonForms/components/Note.tsx:93]: Note, I almost copied this code for something else as well, so probably good to refactor it out at some point...

**Review by clemens-msupply — COMMENTED** [review]
Thanks for the PR think it helps a lot to discuss the design when seeing it in action. 

Regarding your questions.
Think the edit restriction rule is fine (we could remove the age check though so that you can always edit your last message?)

After having a look at the actual UI I am not sure if its a good idea to have the Author as an input field. Might be confusing to the user and better to just record the current user as it is? The use-case was that the current user is not the author of the note but now I wonder if its overkill to record this information give the resulting UX.

Agree, the look of the array could be improved and there is a similar ticket: 
https://github.com/openmsupply/open-msupply/issues/932

Lets discuss in person

**Comment by CarlosNZ** [comment]
> After having a look at the actual UI I am not sure if its a good idea to have the Author as an input field. Might be confusing to the user and better to just record the current user as it is? The use-case was that the current user is not the author of the note but now I wonder if its overkill to record this information give the resulting UX.

My thinking was that we should store the Author's proper name not just their username. Currently the username is the default, but with the possibility of changing it, since we don't currently have the user's full name available.

File: client/packages/programs/src/JsonForms/components/Note.tsx
Comment (mark-prins) [inline client/packages/programs/src/JsonForms/components/Note.tsx:134]: I don't think that the author should be an editable field, you could put anyone's name in here! Makes more sense to tie the author to the logged in user and simply have this as read-only text.

File: client/packages/programs/src/JsonForms/common/components/Array.tsx
Comment (mark-prins) [inline client/packages/programs/src/JsonForms/common/components/Array.tsx:272]: the naming of `authorId` being a bool field is odd to me.
something like `isCurrentUser` makes more sense - I was expecting 'authorId' to be a string prop, not a bool

I don't see the need to be so restrictive - and I would allow a user to edit all of their own notes ( i.e. remove `maxAge` and `latest` )

**Comment by CarlosNZ** [comment]
Okay, @clemens-msupply and @mark-prins, I have made significant improvements, and it looks much nicer now, too:

Screenshot 2023-01-12 at 12 38 21 PM

I'd like to do a bit more tidying up and move some note-specific things into separate files, but I'll wait until I've got approval for the overall logic to prevent the git diff getting out of control.

See additional inline comments for some further explanation.

File: client/packages/programs/src/JsonForms/common/JsonForm.tsx
Comment (CarlosNZ) [inline client/packages/programs/src/JsonForms/common/JsonForm.tsx:31]: I renamed this to avoid a name collision, as I wanted to use a built-in `Array` method, namely: `new Array(x).fill()`

File: client/packages/programs/src/JsonForms/common/components/Array.tsx
Comment (CarlosNZ) [reply client/packages/programs/src/JsonForms/common/components/Array.tsx:315]: I think editability controls for arrays could be useful in a general sense (though by default we never need to worry about them).

File: client/packages/programs/src/JsonForms/common/components/Array.tsx
Comment (CarlosNZ) [inline client/packages/programs/src/JsonForms/common/components/Array.tsx:313]: Some behaviour is Note-specific, so we identify here whether we're dealing with notes or not. As mentioned, I'll pull some of the note-specific code out into another file once we're done here to make this module a little less bloated.

File: client/packages/programs/src/JsonForms/common/components/Array.tsx
Comment (CarlosNZ) [inline client/packages/programs/src/JsonForms/common/components/Array.tsx:376]: Will move to separate file

File: client/packages/programs/src/JsonForms/common/components/Array.tsx
Comment (CarlosNZ) [inline client/packages/programs/src/JsonForms/common/components/Array.tsx:387]: I really wanted to not have to use an absolute pixel value here, but I couldn't get it to work with the overflow/ellipsis functionality. Without an absolute maximum, the text would just expand to the full width of what it *would* be if it weren't truncated. Very annoying. Would love it if we could figure this out.

File: client/packages/programs/src/JsonForms/common/components/Array.tsx
Comment (CarlosNZ) [inline client/packages/programs/src/JsonForms/common/components/Array.tsx:309]: An array of which elements are expanded or aren't (booleans, accessed by index).

These needed to become controlled components so I could know what the current state of each accordion was (since the Notes label is dependent on it)

File: client/packages/programs/src/JsonForms/common/components/Array.tsx
Comment (CarlosNZ) [reply client/packages/programs/src/JsonForms/common/components/Array.tsx:546]: Have added a complete Zod schema for options and validation check in now.

File: client/packages/programs/src/JsonForms/common/components/Array.tsx
Comment (CarlosNZ) [inline client/packages/programs/src/JsonForms/common/components/Array.tsx:66]: Had to make this "any" as there was no way to access existing Type definitions (i.e. "UiSchemaElement"). Should be okay though, as JSONForms itself already validates these elements.

File: client/packages/programs/src/JsonForms/components/Note.tsx
Comment (CarlosNZ) [reply client/packages/programs/src/JsonForms/components/Note.tsx:83]: Yeah, I'll make a new hook for it, since it seems pretty universal.

File: client/packages/programs/src/JsonForms/components/Note.tsx
Comment (CarlosNZ) [reply client/packages/programs/src/JsonForms/components/Note.tsx]: Done in latest changes. 👍

File: client/packages/programs/src/JsonForms/common/components/Array.tsx
Comment (CarlosNZ) [reply client/packages/programs/src/JsonForms/common/components/Array.tsx]: > the naming of `authorId` being a bool field is odd to me.
> something like `isCurrentUser` makes more sense - I was expecting 'authorId' to be a string prop, not a bool

Done :)

> I don't see the need to be so restrictive - and I would allow a user to edit all of their own notes ( i.e. remove `maxAge` and `latest` )

Have left them in for now for demonstration purposes, but I'll pull them out of the uischema before merging (i.e. the options will still exist, but we won't use them (for now). I suspect there will be ongoing discussion about what the "rules" for these are.

File: client/packages/programs/src/JsonForms/components/Note.tsx
Comment (clemens-msupply) [reply client/packages/programs/src/JsonForms/components/Note.tsx:83]: I needed one for the IdGenerator so I created one there already...

File: client/packages/programs/src/JsonForms/common/components/Array.tsx
Comment (clemens-msupply) [inline client/packages/programs/src/JsonForms/common/components/Array.tsx:346]: Think its better to wrap the `return` statement into brackets {} when it spans more than one line. Find it a bit hard to read...

File: client/packages/programs/src/JsonForms/common/components/Array.tsx
Comment (clemens-msupply) [inline client/packages/programs/src/JsonForms/common/components/Array.tsx:555]: Do you remember why this `as` cast was needed?

**Review by clemens-msupply — APPROVED** [review]
Looks good! thank you.

Its still bugging me that you have a Note dependency in the generic Array component though. I believe you can do it like this:

```typescript
const buildArrayComponent = <OptionType extends Options>(
  ZodOption: OptionType,
  getItemLabel: (child: any, index: number, expandable: boolean) => JSX.Element,
  isElementEditable?: (child: any, index: number) => boolean
): ((props: ArrayControlCustomProps) => JSX.Element | null) => {
  return (props: ArrayControlCustomProps) => {
    // content from existing ArrayComponent but using the methods from the arguments
    return null;
  };
};

// Then create the component like this
const ArrayComponent = buildArrayComponent(Options, getItemDefaultLabel, undefined);
// NodeOptions is derived from the default options
const NoteArrayComponent = buildArrayComponent(NodeOptions, getItemNoteLabel, isNoteElementEditable);
```

What do you think?

**Comment by clemens-msupply** [comment]
One more comment, could the edit button go to the top where the remove button and the expand button is? to save some space?

image

but the edit button is in the Note component...

**Comment by CarlosNZ** [comment]
> What do you think?

Yeah, sounds good. I think I'll do this as part of #1015, was planning to do a bit more consolidation there.

File: client/packages/programs/src/JsonForms/common/components/Array.tsx
Comment (CarlosNZ) [reply client/packages/programs/src/JsonForms/common/components/Array.tsx:555]: I *think* it's because the `withJsonFormsArrayControlProps` method is expecting the standard ArrayControlProps, but our versions has a few extra options etc.

**Comment by CarlosNZ** [comment]
> One more comment, could the edit button go to the top where the remove button and the expand button is? to save some space?
> 
> but the edit button is in the Note component...

Yes, exactly, they're in different components. I actually think having a prominent Edit button is okay, since most notes won't be editable. And having a less obvious "Remove" button is also good, since we don't really want to encourage Note removal.

We can re-visit this once a few more people have seen it in action and weighed in?

**Comment by CarlosNZ** [comment]
> Its still bugging me that you have a Note dependency in the generic Array component though. I believe you can do it like this:
> 
> What do you think?

Yeah, that might work, although I did have problems trying to get an additional layer of logic outside the JSONForm components working well with JSONForms (without adding additional complexity). 

Where did you imagine the `buildArrayComponent` would live, and what's it called by? 

I think the only logic here that is *explicitly* only relevant to Notes is the "itemLabel" display, which has the `isNoteArray` check, and some of the internal details of the "edit restrictions". Everything else can be generic to any array type.

I think maybe the main "Array" component can have "Array subTypes", of which we (currently) have `notes` and `generic`, but can add more as required. The options types and any additional logic for `notes` can be moved to another file, and we just have a general `getArraySubtype` method that works out what it is (based on criteria determined elsewhere) and acts accordingly. I can do this (and also maybe try your suggestion) as part of #1015 ?

Either that or I just bite the bullet and make separate components for `Array`, `EnumArray` and `Notes`, and just have a shared "common" file that they each import their common logic from.

Either way, best to do in #1015. Let's discuss before I do that, as [you've raised a couple of other things](https://github.com/openmsupply/open-msupply/issues/1015#issuecomment-1383239564) that need to be worked out too.

**Comment by clemens-msupply** [comment]
> > Its still bugging me that you have a Note dependency in the generic Array component though. I believe you can do it like this:
> > What do you think?
> 
> Yeah, that might work, although I did have problems trying to get an additional layer of logic outside the JSONForm components working well with JSONForms (without adding additional complexity).
> 
> Where did you imagine the `buildArrayComponent` would live, and what's it called by?
> 
> I think the only logic here that is _explicitly_ only relevant to Notes is the "itemLabel" display, which has the `isNoteArray` check, and some of the internal details of the "edit restrictions". Everything else can be generic to any array type.
> 
> I think maybe the main "Array" component can have "Array subTypes", of which we (currently) have `notes` and `generic`, but can add more as required. The options types and any additional logic for `notes` can be moved to another file, and we just have a general `getArraySubtype` method that works out what it is (based on criteria determined elsewhere) and acts accordingly. I can do this (and also maybe try your suggestion) as part of #1015 ?
> 
> Either that or I just bite the bullet and make separate components for `Array`, `EnumArray` and `Notes` (which would include the Array and individual Note code), and just have a shared "common" file that they each import their common logic from.
> 
> Either way, best to do in #1015. Let's discuss before I do that, as [you've raised a couple of other things](https://github.com/openmsupply/open-msupply/issues/1015#issuecomment-1383239564) that need to be worked out too.

Yes if it gets too complex I also think having separate components for Note and Array is probably the best way to do it

**Comment by CarlosNZ** [comment]
@clemens-msupply I've addressed the Array refactoring in #1039. I'll wait until that PR is approved, then I'll merge this both together.

**Pull request merged**
